lib/customisation.nix

```diff
@@ -120,6 +151,64 @@ rec {
       else result);
 
 
+  /* Call the function in the file `fn` with the required
+    arguments automatically.  This should be preferred over callPackage
+    when the result does not need to expose a .override function.  The function
+    is called with the arguments `args`, but any missing arguments are obtained
+    from `autoArgs`.  This function is intended to be partially
+    parameterised, e.g.,
+
+      ```nix
+      callFunction = callFunctionWith pkgs;
+      pkgs = {
+        pkgSet = callFunction ./foo.nix { };
+        helperFn = callFunction ./bar.nix { };
+      };
+      ```
+
+    If the `myPkgSet` function expects an argument named `makeScope`, it is
+    automatically passed as an argument.  Overrides or missing
+    arguments can be supplied in `args`, e.g.
+
+      ```nix
+      python3PkgSet = callFunction ./myPkgSet.nix {
+        newScope = python3.pkgs.newScope;
+      };
+      ```
+
+    <!-- TODO: Apply "Example:" tag to the examples above -->
+
+    Type:
+      callFunctionWith :: AttrSet -> ((AttrSet -> a) | Path) -> AttrSet -> a
+  */
+  callFunctionWith = autoArgs: fn: args:
+    let
+      f = if isFunction fn then fn else import fn;
+      fargs = functionArgs f;
+
+      # All arguments that will be passed to the function
+      # This includes automatic ones and ones passed explicitly
+      allArgs = intersectAttrs fargs autoArgs // args;
+
+      # a list of argument names that the function requires, but
+      # wouldn't be passed to it
+      missingArgs =
+        # Filter out arguments that have a default value
+        (filterAttrs (name: value: ! value)
+        # Filter out arguments that would be passed
+        (removeAttrs fargs (attrNames allArgs)));
+
+      # Partially apply potential arguments
+      errorForArg = errorForFnArgs fn (autoArgs // args) fargs;
+
+      # Only show the error for the first missing argument
+      error = errorForArg (head (attrNames missingArgs));
+
+    in if missingArgs == {}
+       then f allArgs
+       else throw "lib.customisation.callFunctionWith: ${error}";
```

Comment:
The vast majority here seems to be a duplicate to `callFunctionWith`. Since the logic is highly related (calling a function with a set of automatically detected args), wouldn't it make sense to merge the two together?

Reply:
I'll try to refactor it... should be do-able